# Technical Challenge ! Custiomers and Orders - The lifeblood of any business!

This notebook depends on two data files: customers.scv and orders.csv. These files are hosted in the cloud (or Github), so for simplicity, we are providing you with the code to download and save the two files in a folder called 'sample_data' which lives within this notebook session (not on your local computer nor Google Drive).

![sample_data folder in notebook session](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/sample_data_folder.png)

Please execute the first code snippet before moving to the rest of the exercise as this code will import the data from the files for you to be able to execute the rest of the exercise.

Remember as we said in the previous lecture, you have to save this notebook into your Google Drive by going to menu File/Save a Copy In Drive, or clicking in the button "Copy to Drive". This way, you will not lose any work you've done and the file will persist with your latest changes in your Google Drive. Also we recommend you to rename your .ipynb in your Google Drive, so you can easily find it in the future.

In [1]:
import requests

def import_data_files():
  r = requests.get('https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/customers.csv')
  with open('./sample_data/customers.csv', 'wb') as f:
    f.write(r.content)

  r = requests.get('https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/orders.csv')
  with open('./sample_data/orders.csv', 'wb') as f:
    f.write(r.content)
  
import_data_files()
print("Customers and orders CSV files have been added './sample_data'")

Customers and orders CSV files have been added './sample_data'


# Exercise 1: Processing Customers data (difficulty medium)

The sample customer data in 'customers.csv' file has just 5 columns: CustomerId, First Name, Last Name, City and State

![Data sample](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/customers.png)

We strongly recommend that you complete from the [Prep Course: Intro to Python](https://colab.research.google.com/github/anyoneai/notebooks/blob/main/python3_crash_course.ipynb) the following section:
- Section 7: File I/O (to understand how to read CSV file) 
- Section 6: For Loop (to navigate the contents of the CSV file)
- Section 5: Tuples, Lists, and Dictionaries (to manipulate the data of the CSV file)

With this, we hope you can complete this exercise successfully. Although if you want to solve this with libraries or any other way, you are welcome to do it your way.

*Hint:* We advise you to take a look at the data before you start.
**if you want to manually take a look at the data before you start, please see the content of the data [here](https://github.com/anyoneai/notebooks/blob/main/customers_and_orders/data/customers.csv).

*Hint*: There are many ways to do this exercise you can do your own, although here's some help. You can solve this exercise reading and parsing CSV files, structuring data into dictionaries, and using for loops to navigate the contents

*Hint*: Also, keep in mind that data might not be clean and you might have to figure out how to deal with that data from the code, without having to modify the data source.

Below are the 5 questions you'll have to answer to pass the evaluation:

**Libraries**

In [2]:
#Importing libraries
import numpy as np
import pandas as pd
import scipy
import os 

**Import** **Data**

In [7]:
#Importing data
data = pd.read_csv('./sample_data/customers.csv')
data.head(1)

CustomerID FirstName    LastName         City State
0        6192     Randi  Piedrahita  Canoga Park    CA

**Explore Data**

In [8]:
#Types of data
data.dtypes

CustomerID     int64
FirstName     object
LastName      object
City          object
State         object
dtype: object

In [9]:
#Check the data

data.isnull()
data.isnull().sum()

CustomerID    0
FirstName     0
LastName      0
City          0
State         0
dtype: int64

In [10]:
#checking the distribution 
data.count()

CustomerID    602
FirstName     602
LastName      602
City          602
State         602
dtype: int64

**Question 1:** How many customers are in the file?
(as help, we have added some comments and starter code to help you structure the solution)

In [11]:
from os.path import exists
import csv

# check if file can be found and open the file
datafile = "./sample_data/customers.csv"
if not exists(datafile):
  raise SystemExit("You should run the first code cell and download the dataset files!")

# customers file is in the sample_data folder and can be opened.
# Using csv reader is one way of dealing with the file 
# but it's up to you how you want to proceed
with open(datafile, 'r') as fl:
  csvreader = csv.reader(fl, delimiter=',')

  # loop through the rows and count etc. THE REST IS UP TO YOU!


print("All done!")

All done!


In [12]:
data.shape

(602, 5)

In [13]:
data.info

<bound method DataFrame.info of      CustomerID    FirstName      LastName         City State
0          6192        Randi    Piedrahita  Canoga Park    CA
1       5100595  Christopher       Abraham   Loma Linda    CA
2       1902451        Elvis      Jauregui  Los Angeles    CA
3       1906377         Jose  Mota-Sanchez       Downey    CA
4       1902406      Scott        McCrorie   Long Beach    CA
..          ...          ...           ...          ...   ...
597     5415365      Trinity     Hall Reif        Poway    CA
598     5415465       Kisona   Mialoto Jr.      Lynwood    CA
599     5415345     Lawrence         Zhang  Temple City    CA
600     5415481         John        Hughes    Round Top    TX
601     5415482      Phillip        Okonma  Los Angeles    CA

[602 rows x 5 columns]>

Answer: There are 602 Customers

**Question 2:** In how many different states do the customers live in?

In [ ]:
#Importing data

In [ ]:
# open file and init vars
# evaluate the contents of the file

In [14]:
data['State'].unique()

array(['CA', 'AZ', 'NV', 'FL', 'CA           ', 'WA', 'NH', 'ID', 'CO',
       'Ca', 'TX', 'NM', 'OR', 'UT', 'MA', 'IN'], dtype=object)

In [15]:
data['State'].unique().tolist()

['CA',
 'AZ',
 'NV',
 'FL',
 'CA           ',
 'WA',
 'NH',
 'ID',
 'CO',
 'Ca',
 'TX',
 'NM',
 'OR',
 'UT',
 'MA',
 'IN']

In [16]:
len(data['State'].unique())

16

Answer: They live in 14 different states, (CA, Ca and "CA " are the same state. Will clean this later)

**Question 3** What is the state with most customers?

As you can see, There are 3 diferentes names for CA state. We will fix it.

In [17]:
data['State'] =  data['State'].replace({'Ca':'CA'})
data['State'] =  data['State'].replace({'CA           ':'CA'})

In [18]:
data['State'].unique().tolist()

['CA',
 'AZ',
 'NV',
 'FL',
 'WA',
 'NH',
 'ID',
 'CO',
 'TX',
 'NM',
 'OR',
 'UT',
 'MA',
 'IN']

In [19]:
group=data.groupby("State")
print(group.size().reset_index(name='counts'))

   State  counts
0     AZ       6
1     CA     569
2     CO       3
3     FL       3
4     ID       1
5     IN       1
6     MA       1
7     NH       1
8     NM       3
9     NV       8
10    OR       1
11    TX       2
12    UT       2
13    WA       1


Answer: The state with the most number of customers, is CA with 569 customers

**Question 4** What is the state with the least customers?

Answer: using the code displayed for Question 3, we can see that there are 6 States with 1 customer: ID,IN,MA,NH,OR,WA.

**Question 5:** What is the most common last name?

In [20]:
lista = pd.Series(data["LastName"])
resultados = lista.value_counts()

print(resultados)

Smith       8
Zambrana    5
Gomez       5
Doggett     5
Rocha       4
           ..
Bagale      1
Puno        1
Kovelman    1
Vrbka       1
Okonma      1
Name: LastName, Length: 424, dtype: int64


Answer: The most common last name is Smith.

# Exercise 2: Processing Orders data (difficulty high)

The second sample files contains orders placed by customers from the first file. Be careful, this file has many rows and you most likely should not print the contents of the whole file. 

The file contains the following columns: CustomerID, OrderID, Date, OrderTotal, ProductName, Price

![Data sample](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/orders.png)

*Hint:* We advise you to take a look at the data before you start.
**if you want to manually take a look at the data before you start, please see the content of the data [here](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/orders.csv).

*Hint*: There are many ways to do this exercise you can do your own, although here's some help. You can solve this exercise reading and parsing CSV files, structuring data into dictionaries, and using for loops to navigate the contents

*Hint*: Also, the data is not clean and you will have to figure out how to deal with that data from the code, without having to modify the data source.



**Question #1:** How many unique orders are in the orders.csv file?

**Question #2:** What is the average number of items per order (rounded to two decimal places)?

**Question #3:** What is the highest number of items per order?

**Question #4:** What is the number of orders placed in October 2021?

**Question #5:** Which customer spent the most amount of money in 2021?

**Question #6:** Historically, what is the best month for sales?

Once you get your answers, remember to go back to the course and introduce them in the multiple choice quiz

In [28]:
#Importing data
orders = pd.read_csv('./sample_data/orders.csv')
data.head(5)

CustomerID  OrderID                     Date  OrderTotal ProductName  Price
0        8091  7742581  2021-07-26 14:40:10.783        95.0         Z03   90.0
1      902139  7742778  2021-08-08 05:01:21.120        60.0    0844 A/C   60.0
2     2300266  7742593  2021-07-27 11:00:16.020       185.0         M07   90.0
3     2300266  7742593  2021-07-27 11:00:16.020       185.0         M09   90.0
4     5173013  7742609  2021-07-28 14:26:13.930       165.0        0324  160.0

**Explore Data**

In [29]:
#Types of data
orders.dtypes

CustomerID       int64
OrderID          int64
Date            object
OrderTotal     float64
ProductName     object
Price          float64
dtype: object

In [32]:
#Check the data

orders.isnull()
orders.isnull().sum()

CustomerID        0
OrderID           0
Date           2142
OrderTotal        0
ProductName       0
Price             0
dtype: int64

In [33]:
#checking the distribution 
orders.count()

CustomerID     29294
OrderID        29294
Date           27152
OrderTotal     29294
ProductName    29294
Price          29294
dtype: int64

**Question #1:** How many unique orders are in the orders.csv file?

In [34]:
orders.shape

(29294, 6)

In [35]:
len(orders['OrderID'].unique())

16672

Answer: There are 16672 unique orders in the dataset

**Question #2:** What is the average number of items per order (rounded to two decimal places)?

In [45]:
Raport = orders.groupby(['OrderID']).count()
print (Raport)
OrderTotal = Raport['OrderTotal']
print(OrderTotal)

         CustomerID  Date  OrderTotal  ProductName  Price
OrderID                                                  
5038              1     0           1            1      1
5052              2     0           2            2      2
5058              2     0           2            2      2
5126              1     0           1            1      1
5130              2     0           2            2      2
...             ...   ...         ...          ...    ...
7791846           1     1           1            1      1
7791847           1     1           1            1      1
7791848           1     1           1            1      1
7791849           3     3           3            3      3
7791850           3     3           3            3      3

[16672 rows x 5 columns]
OrderID
5038       1
5052       2
5058       2
5126       1
5130       2
          ..
7791846    1
7791847    1
7791848    1
7791849    3
7791850    3
Name: OrderTotal, Length: 16672, dtype: int64


In [40]:
#checking an easier, clever way
Report = orders.groupby(by =["OrderID"])['ProductName'].count()
print(Report)

OrderID
5038       1
5052       2
5058       2
5126       1
5130       2
          ..
7791846    1
7791847    1
7791848    1
7791849    3
7791850    3
Name: ProductName, Length: 16672, dtype: int64


In [46]:
#Calculation the mean o avg item per order
from statistics import mean
round(mean(Report),2)

1.76

In [47]:
#Calculation the mean o avg item per order (with first calculation approach)
from statistics import mean
round(mean(OrderTotal),2)

1.76

Answer: The avg items per order is 1.76

**Question #3:** What is the highest number of items per order?

In [48]:
max(Report)

35

In [49]:
max(OrderTotal)

35

In [51]:
Highest = (orders.groupby(["OrderID"])["ProductName"]
           .count().sort_values(ascending=False)).to_frame()

print(Highest)

         ProductName
OrderID             
5802666           35
7411097           30
7553706           30
7001632           30
7005266           28
...              ...
4217400            1
4217390            1
4217389            1
4217387            1
5038               1

[16672 rows x 1 columns]


Answer: The highest number of items per order is 35

**Question #4:** What is the number of orders placed in October 2021?

In [55]:
import datetime
orders["Orddate"] = pd.to_datetime(orders["Date"])

orders.head(2)

CustomerID  OrderID                     Date  OrderTotal ProductName  \
0        8091  7742581  2021-07-26 14:40:10.783        95.0         Z03   
1      902139  7742778  2021-08-08 05:01:21.120        60.0    0844 A/C   

   Price                 Orddate  
0   90.0 2021-07-26 14:40:10.783  
1   60.0 2021-08-08 05:01:21.120

In [56]:
orders.dtypes

CustomerID              int64
OrderID                 int64
Date                   object
OrderTotal            float64
ProductName            object
Price                 float64
Orddate        datetime64[ns]
dtype: object

In [57]:
orders['Month'] = orders["Orddate"].dt.month
orders['Year'] = orders["Orddate"].dt.year
orders.head(2)

CustomerID  OrderID                     Date  OrderTotal ProductName  \
0        8091  7742581  2021-07-26 14:40:10.783        95.0         Z03   
1      902139  7742778  2021-08-08 05:01:21.120        60.0    0844 A/C   

   Price                 Orddate  Month    Year  
0   90.0 2021-07-26 14:40:10.783    7.0  2021.0  
1   60.0 2021-08-08 05:01:21.120    8.0  2021.0

In [60]:
orders.groupby(["Year","Month"]).size().reset_index().tail(20)

Year  Month    0
137  2019.0    8.0  256
138  2019.0    9.0  256
139  2019.0   10.0  265
140  2019.0   11.0  302
141  2019.0   12.0  217
142  2020.0    1.0  338
143  2020.0    2.0  302
144  2020.0    3.0  217
145  2021.0    3.0  104
146  2021.0    4.0  308
147  2021.0    5.0  410
148  2021.0    6.0  384
149  2021.0    7.0  407
150  2021.0    8.0  379
151  2021.0    9.0  372
152  2021.0   10.0  437
153  2021.0   11.0  339
154  2021.0   12.0  355
155  2022.0    1.0  670
156  2022.0    2.0  156

Answer: The number of orders placed in October 2021 is 437

**Question #5:** Which customer spent the most amount of money in 2021?

In [62]:
#We will convert the feature "OrderTotal" to numeric
orders ['OrderTotal'] = orders[['OrderTotal']].astype(float)

In [63]:
orders.dtypes

CustomerID              int64
OrderID                 int64
Date                   object
OrderTotal            float64
ProductName            object
Price                 float64
Orddate        datetime64[ns]
Month                 float64
Year                  float64
dtype: object

In [64]:
orders.head()

CustomerID  OrderID                     Date  OrderTotal ProductName  \
0        8091  7742581  2021-07-26 14:40:10.783        95.0         Z03   
1      902139  7742778  2021-08-08 05:01:21.120        60.0    0844 A/C   
2     2300266  7742593  2021-07-27 11:00:16.020       185.0         M07   
3     2300266  7742593  2021-07-27 11:00:16.020       185.0         M09   
4     5173013  7742609  2021-07-28 14:26:13.930       165.0        0324   

   Price                 Orddate  Month    Year  
0   90.0 2021-07-26 14:40:10.783    7.0  2021.0  
1   60.0 2021-08-08 05:01:21.120    8.0  2021.0  
2   90.0 2021-07-27 11:00:16.020    7.0  2021.0  
3   90.0 2021-07-27 11:00:16.020    7.0  2021.0  
4  160.0 2021-07-28 14:26:13.930    7.0  2021.0

In [66]:
#We will group the total with each customer
order_tot = orders.groupby(['CustomerID','Year']).sum()["OrderTotal"].to_frame()
order_tot.head(10)

OrderTotal
CustomerID Year              
5014       2008.0      1720.0
           2009.0      6550.0
           2010.0      6490.0
           2011.0      6330.0
           2012.0      6770.0
           2013.0      6970.0
           2014.0      7850.0
           2015.0      6870.0
           2016.0     11065.0
           2017.0     12590.0

In [68]:
order_tot = order_tot.reset_index()
order_tot.head(10)

index  CustomerID    Year  OrderTotal
0      0        5014  2008.0      1720.0
1      1        5014  2009.0      6550.0
2      2        5014  2010.0      6490.0
3      3        5014  2011.0      6330.0
4      4        5014  2012.0      6770.0
5      5        5014  2013.0      6970.0
6      6        5014  2014.0      7850.0
7      7        5014  2015.0      6870.0
8      8        5014  2016.0     11065.0
9      9        5014  2017.0     12590.0

In [69]:
order_tot[(order_tot['Year']==2021.0)].max()

index            2333.0
CustomerID    5415350.0
Year             2021.0
OrderTotal      56170.0
dtype: float64

Answer: The customer who spent the most amount of money in 2021 was "CustomerID 5415350.0" with $56.170

**Question #6:** Historically, what is the best month for sales?

In [72]:
order_tot = orders.groupby(['Year','Month']).sum()["OrderTotal"].to_frame()
order_tot.head()

OrderTotal
Year   Month            
2005.0 10.0        490.0
2006.0 11.0        375.0
2007.0 3.0           5.0
2008.0 6.0          60.0
       7.0        4335.0

In [74]:
order_tot2 = order_tot.reset_index()
order_tot2.head(20)

Year  Month  OrderTotal
0   2005.0   10.0       490.0
1   2006.0   11.0       375.0
2   2007.0    3.0         5.0
3   2008.0    6.0        60.0
4   2008.0    7.0      4335.0
5   2008.0    8.0      3350.0
6   2008.0    9.0      1190.0
7   2008.0   10.0     18200.0
8   2008.0   11.0     27200.0
9   2008.0   12.0     15355.0
10  2009.0    1.0     36350.0
11  2009.0    2.0     43960.0
12  2009.0    3.0     16095.0
13  2009.0    4.0     25030.0
14  2009.0    5.0     34335.0
15  2009.0    6.0     45020.0
16  2009.0    7.0     24000.0
17  2009.0    8.0     36315.0
18  2009.0    9.0     19320.0
19  2009.0   10.0     29910.0

In [73]:
idx = order_tot2[(order_tot2['Year']> 2005.0)]['OrderTotal'].idxmax()
df_max = order_tot2.loc[idx,]
df_max

Year            2022.0
Month              1.0
OrderTotal    172032.0
Name: 155, dtype: float64

Answer: Historically, the best month was January of the year 2022

# Finished!

Hope this was not too difficult and slicing and dicing the datasets was some fun. Now head on back to the course and provide the answers to the questions from this exercise.